Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds =
+                    options.SlidingExpiration.Value.TotalSeconds;
+            }
+            else
+            {
+                var forever = DateTime.MaxValue - doc.LastRefresh.Value;
+                doc.SlidingExpirationSeconds = forever.TotalSeconds / 2;
+            }
+            if (options.AbsoluteExpiration.HasValue)
+            {
+                doc.AbsoluteExpiration = options.AbsoluteExpiration.Value.UtcDateTime;
+            }
+            if (options.AbsoluteExpirationRelativeToNow.HasValue)
+            {
+                doc.AbsoluteExpiration = DateTime.UtcNow +
+                    options.AbsoluteExpirationRelativeToNow.Value;
+            }
+            return doc;
+        }
+
+        void IDistributedCache.Set(string key, byte[] value, DistributedCacheEntryOptions options) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options)).Wait();
+
+        Task IDistributedCache.SetAsync(string key, byte[] value, DistributedCacheEntryOptions options,
+            CancellationToken token) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options),
+            cancellationToken: token);
+
+        /// <summary>
+        /// Scans the firestore collection for expired cache entries and
+        /// deletes them.  
+        /// </summary>
+        /// <param name="token">A cancellation token.</param>
+        public async Task CollectGarbageAsync(CancellationToken token)
+        {
+            _logger.LogTrace("Begin garbage collection.");
+            // Purge entries whose AbsoluteExpiration has passed.
+            const int pageSize = 40;
+            int batchSize;  // Batches of cache entries to be deleted.
+            var now = DateTime.UtcNow;
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderByDescending("AbsoluteExpiration")
+                    .StartAfter(now)
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    if (docSnapshot.ConvertTo<CacheDoc>()
+                        .AbsoluteExpiration.HasValue)
+                    {
+                        writeBatch.Delete(docSnapshot.Reference,
+                            Precondition.LastUpdated(
+                                docSnapshot.UpdateTime.GetValueOrDefault()));
+                        batchSize += 1;
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);

```

Comment:
Suggest changing `{0}` to `{batchSize}` for clarity in the logs.

Reply:
My suspicion is that the logging infrastructure won't even compose the log string if it never gets written anywhere, so this code is more efficient in that case.

Reply:
No, I'm not suggesting making it an interpolated string literal - I'm suggesting it should just be

```csharp
_logger.LogDebug("Collecting {batchSize} cache entries.", batchSize);
```

I know it *looks* like it's a format string, but it's not - it surprised me too!